### PR TITLE
Add support for determining encoding from multi value headers

### PR DESCRIPTION
**Pull request by astrikos**
## Which problem is this PR solving?
Some applications (kubernetes apiserver in my case) might send headers like ``application/json, */*`` which are legit according to RFC 26116 section 4.2 (https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2).

The current version doesn't take that into account and is returning an error in the` determineEncoding` [function](https://github.com/beatlabs/patron/blob/master/component/http/handler.go#L53) instead of the json which would be expected in the above example.

## Short description of the changes
Change `determineEncoding` [function](https://github.com/beatlabs/patron/blob/master/component/http/handler.go#L53) to parse headers and compare all available values of the header to the needed values.

Addresses https://github.com/beatlabs/patron/issues/215

### CODE REVIEW

**Review by drakos74 — DISMISSED** [review]
Good addition ! I am just thinking if we should fail in the case where we have contradicting headers amongs the multi values

File: component/http/handler.go [hunk 1]
```diff
@@ -62,44 +62,76 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
...
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			dec = protobuf.Decode
-			ct = protobuf.Type
-		default:
+		foundValidHeader = false
+		ah := getMultiValueHeaders(cth[0])
+		for _, v := range ah {
+			switch v {
+			case "*/*", json.Type, json.TypeCharset:
```
Comment (mantzas) [inline component/http/handler.go]: With the above for loop the last match wins. Maybe we should break when the first match happens.

File: component/http/handler.go [hunk 2]
```diff
@@ -62,44 +62,76 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
...
-			ct = json.TypeCharset
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			if dec == nil {
-				dec = protobuf.Decode
+		foundValidHeader = false
+		ah := getMultiValueHeaders(ach[0])
+		for _, v := range ah {
+			switch v {
+			case "*/*", json.Type, json.TypeCharset:
```
Comment (mantzas) [inline component/http/handler.go]: With the above for loop the last match wins. Maybe we should break when the first match happens.

**Review by mantzas — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution.
Would it be possible to open an issue and link the PR to this issue?
(Contribution Guidelines)

**Comment by astrikos** [comment]
>I am just thinking if we should fail in the case where we have contradicting headers amongst the multi values

Yeah maybe for content headers would make sense since I guess it can send either xml or json.
But for the accept header it could be both, since that would mean that the server can accept both, which is valid, right?

I could write more logic for the content header but I don't know if it's worth spending so much time on checking the headers so thoroughly.
It's better to be liberal in what you accept according to [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle) :) but I am okay adding more logic if you guys think that's the best here.

File: component/http/handler.go (hunk 1 above)
Comment (astrikos) [reply component/http/handler.go]: Alright I pushed a fix but I am not fond of named loops, I could also get rid of switch and have an if/else if now that we don't have the default anymore. Whatever you like more I don't have strong feelings.

**Comment by astrikos** [comment]
I see tests now are broken on kafka tests, which are unrelated to my change :thinking: , I'll try to get the latest master and see if that fixes it.

Also new issue created and linked in here https://github.com/beatlabs/patron/issues/215

**Comment by mantzas** [comment]
> I see tests now are broken on kafka tests, which are unrelated to my change thinking , I'll try to get the latest master and see if that fixes it.
> 
> Also new issue created and linked in here #215

Yes, it is already known and I am already working on it in #212 so no need for a new issue.

File: component/http/handler.go (hunk 1 above)
Comment (mantzas) [reply component/http/handler.go]: I also do not like of labels.  You can actually get rid of them just by putting this code inside a new function and return the values set...

Comment (astrikos) [reply component/http/handler.go]: Sounds good, will try that

File: component/http/handler.go [hunk 3]
```diff
@@ -62,44 +62,77 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
...
+		ah := getMultiValueHeaders(cth[0])
+		for _, v := range ah {
+			ct, dec, enc, err = getSingleHeaderEncoding(v)
+			if err == nil {
+				break
+			}
+		}
+		// No valid headers found after going through all headers
+		if err != nil {
+			return "", nil, nil, errors.New("accept header not supported")
```
Comment (mantzas) [inline component/http/handler.go]:
```suggestion
			return "", nil, nil, errors.New("content type header not supported")
```

File: component/http/handler.go [hunk 4]
```diff
@@ -62,44 +62,77 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
...
-			ct = json.TypeCharset
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			dec = protobuf.Decode
-			ct = protobuf.Type
-		default:
-			return "", nil, nil, errors.New("content type header not supported")
+		ah := getMultiValueHeaders(cth[0])
+		for _, v := range ah {
+			ct, dec, enc, err = getSingleHeaderEncoding(v)
```
Comment (mantzas) [inline component/http/handler.go]: Question: Let's assume that we have 3 encodings and the 3rd is correct but the previous 2 are not supported, should we fail? Maybe we need just a correct one.

File: component/http/handler.go [hunk 5]
```diff
@@ -62,44 +62,77 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
...
-			enc = json.Encode
-			dec = json.Decode
-			ct = json.TypeCharset
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			dec = protobuf.Decode
-			ct = protobuf.Type
-		default:
-			return "", nil, nil, errors.New("content type header not supported")
+		ah := getMultiValueHeaders(cth[0])
```
Comment (mantzas) [inline component/http/handler.go]: Question: Will there be ever more than one content-type? When i am POSTing a request it cannot be both JSON and XML. Maybe this applies only for the accept header...

BTW: the variable ah should be ch for content header.

File: component/http/handler.go [hunk 6]
```diff
@@ -62,44 +62,77 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
...
+	default:
+		return "", nil, nil, errors.New("accept header not supported")
+	}
+
+	return ct, dec, enc, nil
+}
+
+func getMultiValueHeaders(header string) []string {
+	if !strings.Contains(header, ",") {
+		return []string{header}
```
Comment (mantzas) [inline component/http/handler.go:130]: We could not check for Contains and do split directly. Would that work?

**Review by mantzas — CHANGES_REQUESTED** [review]

File: component/http/handler.go (hunk 5 above)
Comment (astrikos) [reply component/http/handler.go]: yeah maybe it doesn't make sense for content, I think I added for consistency but it doesn't make a lot of logical sense. 
Alright remove it, added the old code and adjust the logic only for the accept headers.

File: component/http/handler.go (hunk 6 above)
Comment (astrikos) [reply component/http/handler.go:130]: oh you mean skip `Contains` and call `Split` directly? something like:
```
func getMultiValueHeaders(header string) []string {
  splitHeaders := strings.Split(header, ",")
  if splitHeaders[0] == header {
    return []string{header}
  }
  ....
```

Comment (mantzas) [reply component/http/handler.go:130]: yes, the cost of doing split vs contains might the same...maybe a benchmark could shed light if we want to be data-driven...
benchmark with single and multi heder with and without contains...

Comment (drakos74) [reply component/http/handler.go:130]: i did some benchmark on that ...
```
GOROOT=/usr/local/go #gosetup
GOPATH=/Users/vangeliskatikaridis/go #gosetup
/usr/local/go/bin/go test -c -o /private/var/folders/f1/w8ll0_cj1bl_prh_f928f0zr0000gp/T/___gobench_github_com_beatlabs_patron_component_http github.com/beatlabs/patron/component/http #gosetup
/private/var/folders/f1/w8ll0_cj1bl_prh_f928f0zr0000gp/T/___gobench_github_com_beatlabs_patron_component_http -test.v -test.bench . -test.run ^$ #gosetup
goos: darwin
goarch: amd64
pkg: github.com/beatlabs/patron/component/http
BenchmarkSingleHeaderContains
BenchmarkSingleHeaderContains-16    	34828395	        30.6 ns/op
BenchmarkMultiHeaderContains
BenchmarkMultiHeaderContains-16     	 6916483	       173 ns/op
BenchmarkSingleHeaderSplit
BenchmarkSingleHeaderSplit-16       	16266246	        72.5 ns/op
BenchmarkMultiHeaderSplit
BenchmarkMultiHeaderSplit-16        	 7079029	       167 ns/op
PASS
```

Seems in the case of single value headers, the performance is better (by a factor of around 2) . 
For the multi value case , as probably expected, the performance is very much comparable.
In absolute values, the gain of using Contains for single headers is around 40ns

@astrikos , @mantzas  . What do you think ?

attaching my test class as reference

```
package http

import (
	"gotest.tools/assert"
	"strings"
	"testing"
)

func getMultiValueHeadersContains(header string) []string {
	if !strings.Contains(header, ",") {
		return []string{header}
	}

	splitHeaders := strings.Split(header, ",")

	trimmedHeaders := []string{}
	for _, v := range splitHeaders {
		trimmedHeaders = append(trimmedHeaders, strings.TrimSpace(v))
	}

	return trimmedHeaders
}

var headers []string

func getMultiValueHeadersSplit(header string) []string {
	splitHeaders := strings.Split(header, ",")

	trimmedHeaders := []string{}
	for _, v := range splitHeaders {
		trimmedHeaders = append(trimmedHeaders, strings.TrimSpace(v))
	}

	return trimmedHeaders
}

func TestSingleValueHeadersContains(t *testing.T) {

	headers := getMultiValueHeadersContains("string")

	assert.DeepEqual(t,[]string{"string"},headers)

}

func TestMultiValueHeadersContains(t *testing.T) {

	headers := getMultiValueHeadersContains("string,string")

	assert.DeepEqual(t,[]string{"string","string"},headers)

}

func TestSingleValueHeadersSplit(t *testing.T) {

	headers := getMultiValueHeadersContains("string")

	assert.DeepEqual(t,[]string{"string"},headers)

}

func TestMultiValueHeadersSplit(t *testing.T) {

	headers := getMultiValueHeadersContains("string,string")

	assert.DeepEqual(t,[]string{"string","string"},headers)

}

func BenchmarkSingleHeaderContains(b *testing.B){

	benchmarkMultiValueHeadersContains("string",b)

}

func BenchmarkMultiHeaderContains(b *testing.B){

	benchmarkMultiValueHeadersContains("string,string",b)

}


func BenchmarkSingleHeaderSplit(b *testing.B){

	benchmarkMultiValueHeadersSplit("string",b)

}

func BenchmarkMultiHeaderSplit(b *testing.B){

	benchmarkMultiValueHeadersSplit("string,string",b)

}

func benchmarkMultiValueHeadersContains(header string, b *testing.B) {
	for n := 0; n < b.N; n++ {
		headers = getMultiValueHeadersContains(header)
	}
}

func benchmarkMultiValueHeadersSplit(header string, b *testing.B) {
	for n := 0; n < b.N; n++ {
		headers = getMultiValueHeadersSplit(header)
	}
}

```

Comment (astrikos) [reply component/http/handler.go:130]: I am happy with `Contains` for this case to be honest, it's comparable and I think a bit more readable. 
@mantzas if you are okay with that I am okay to move on and merge that.

**Comment by astrikos** [comment]
@mantzas @drakos74 rebased from master, tests looks good locally and on ci, let me know if you need anything more before we can merge this one.

**Review by mantzas — APPROVED** [review]

**Review by tpaschalis — APPROVED** [review]
Good job, thank you for following through and contributing a helpful feature!

**Review by drakos74 — APPROVED** [review]

**Pull request merged**